### DISCUSSION TITLE
Detecting issue with Font in a PDF

### DISCUSSION THREAD

**Discussion by qwertynik** [pymupdf/PyMuPDF#1596] [Looking for help]
@JorjMcKie 

The below snippet of code was shared by you earlier. While this manages to extract the font for most cases, there are cases when the text rendered in a PDF using the extracted fonts is jumbled.

```python
fitz.TOOLS.set_subset_fontnames(True)  # return subset fontnames
fontitems = page.get_fonts()  # list of all fonts
fonts = {}  # fontname -> fitz.Font
for f in fontitems:
    if f[1] == "n/a":  # not embedded: Helvetiva -> fitz.Font("Helvetica")
        if f[2] == "Type3":
            raise RuntimeError("cannot handle Type 3 fonts")
        fonts[f[3]] = fitz.Font(f[3])
    else:  # extract the buffer and make a font of it
        fname = f[3]
        buffer = doc.extract_font(f[0])[-1]
        font = fitz.Font(fontbuffer=buffer)
        fonts[fname] = font
```

Any ideas to detect errors in the font extraction?

**Comment by JorjMcKie** [comment]
I would need an example.
There are cases, when the fontname returned by `get_text("dict")` is hidden somewhere in the PDF font definition, i.e. several connected xrefs jointly describe one font, but may contain (slightly) different names for that same font.
If you look at the details of the font replacing scripts, you will find code that tries to trace this situation - should be in `repl-fontnames.py`.

**Reply by qwertynik** [reply]
Had an example, but do not have access to it now. Will check the `repl-fontnames.py` script for more ideas on this and revisit this discussion.

**Reply by qwertynik** [reply]
@JorjMcKie 
Went through the script `repl-fontnames.py`, did not find anything in particular that attempts to detect font errors in a PDF.

**Comment by JorjMcKie** [comment]
If this happens, then already the text extraction from the source page delivered invalid characters `0xFFFD`. This may happen when the font is unsupported by MuPDF, or contains inbuilt protection against extraction, etc.
Except for "punctual" OCRing text spans containing those character codes (there is a script) I see no way out.
The font replacer scripts indeed contain no precaution for this ...

**Reply by qwertynik** [reply]
> If this happens, then already the text extraction from the source page delivered invalid characters `0xFFFD`

Sure @JorjMcKie, will look for indications. This becomes tricky when PDF itself contains text `0xFFFD` though - any ideas on tackling this?

>This may happen when the font is unsupported by MuPDF, or contains inbuilt protection against extraction, etc.

Are there ways using which the above scenarios be detected?

>Except for "punctual" OCRing text spans containing those character codes

Can more info on what "punctual" OCR means be added?

**Reply by JorjMcKie** [reply]
Look at [this](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/OCR/tesseract2.py) script.
It walks through the dict `page.get_text("dict",...)` and OCRs a span if it contains at least one `chr(65533)`. The OCRed text of that span replaces the original unreadable one.

**Reply by JorjMcKie** [reply]
> when PDF itself contains text `0xFFFD` though

This probably **_never_** is the case. The creating software certainly **_does_** know the unicode character it wants to output, and it also knows which glyph number to hand over to the font.
The font in turn has painting instructions for each glyph which create the respective graphical appearance (which is a combination of line and Beziér curve drawings).
PDF consumers (i.e. viewers, including MuPDF) read the glyph number from the file and look up the corresponding painting instructions to create the visual appearance. This part is bound to always work. But **_if the font does not exhibit_** the original unicode (and it is not obliged to do that) for that glyph, the consumer cannot know it obviously, and will return `0xFFFD` (plus the glyph number in case of `page.get_texttrace()`) in text extractions.
In these cases, only OCR can generate the original unicode.

**Reply by qwertynik** [reply]
The interoperation between the font and PDF read/write software is clearer now @JorjMcKie. Thanks!
There could still be cases, probably rare, where the PDF itself contains the �(0xFFFD) character. And in these cases, there would be no way to ascertain if the font present in the PDF has issues.

**Reply by JorjMcKie** [reply]
> where the PDF itself contains the �(0xFFFD) character

That is correct - and may be intentional. You can output text containing this character yourself with PyMuPDF. Detecting real errors in a font is not (Py-) MuPDF's business however, and in most cases not distinguishable from the other mentioned situations. There is font software specialized on these things.
This also is of no interest for your use case if I understood well.

**Reply by JorjMcKie** [reply]
Also, the font may be or have been okay before insertion, but some of its component may have been deleted / damaged in the PDF due to some other error.
Create this type of situation like so:
```python
import fitz
doc=fitz.open()
page=doc.new_page()
tw=fitz.TextWriter(page.rect)
tw.append((100,100),"just some text")
tw.write_text(page)
print(page.get_fonts())
[(5, 'cid', 'Type0', 'Helvetica', 'F0', 'Identity-H')]
doc.xref_set_key(5, "ToUnicode", "null")  # remove glyph-to-unicode mapping
doc.save("x.pdf")
doc.close()
doc=fitz.open("x.pdf")
page=doc[0]
print(page.get_text())
��������������
```
But the PDF is still viewable nicely!

**Reply by qwertynik** [reply]
A great example @JorjMcKie. Makes it all the more tricky. But also shows the robustness or ambiguity (depending upon how one interacts with the PDF 😂) of the PDF specification

**Reply by JorjMcKie** [reply]
Now you know what to do if you want to immunize your text against extraction - without distroying the appearance (just kidding, there arre better ways!).

**Reply by qwertynik** [reply]
Haha, yeah! Would have used this approach to make extraction difficult. But since you mentioned there are better ways, will look for them when required.
